src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py

```diff
@@ -49,20 +42,42 @@ class SubvolumeV2(SubvolumeBase, SubvolumeTemplate):
     def version():
         return SubvolumeV2.VERSION
 
-    @property
-    def path(self):
-        try:
-            # no need to stat the path -- open() does that
-            return self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_PATH).encode('utf-8')
-        except MetadataMgrException as me:
-            raise VolumeException(-errno.EINVAL, "error fetching subvolume metadata")
-
     @property
     def features(self):
         return [SubvolumeFeatures.FEATURE_SNAPSHOT_CLONE.value,
                 SubvolumeFeatures.FEATURE_SNAPSHOT_AUTOPROTECT.value,
                 SubvolumeFeatures.FEATURE_SNAPSHOT_RETENTION.value]
 
+    @staticmethod
+    def is_valid_uuid(uuid_str):
+        try:
+            uuid.UUID(uuid_str)
+            return True
+        except ValueError:
+            return False
+
+    def snapshot_base_path(self):
+        return os.path.join(self.base_path, self.vol_spec.snapshot_dir_prefix.encode('utf-8'))
+
+    def snapshot_data_path(self, snapname):
+        snap_base_path = self.snapshot_path(snapname)
+        uuid_str = None
+        with self.fs.opendir(snap_base_path) as dir_handle:
```

Comment:
snapshot info on non-existing snapshot crashes here.

```
# bin/ceph fs subvolume snapshot info a sub_0 snap_1
Error EINVAL: Traceback (most recent call last):
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/mgr_module.py", line 1170, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/module.py", line 357, in handle_command
    return handler(inbuf, cmd)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/module.py", line 471, in _cmd_fs_subvolume_snapshot_info
    group_name=cmd.get('group_name', None))
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/fs/volume.py", line 345, in subvolume_snapshot_info
    snap_info_dict = subvolume.snapshot_info(snapname)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py", line 273, in snapshot_info
    snappath = self.snapshot_data_path(snapname)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py", line 65, in snapshot_data_path
    with self.fs.opendir(snap_base_path) as dir_handle:
  File "cephfs.pyx", line 1073, in cephfs.LibCephFS.opendir
cephfs.ObjectNotFound: opendir failed: No such file or directory [Errno 2]

```

Reply:
Good catch, will add a test case and also add exception handling within snapshot_data_path to only throw VolumeException on errors.

Reply:
Fixed.